### PR TITLE
Use Task class instead of tuple

### PR DESCRIPTION
**Pull request by fjetter**
This is an early version that will close https://github.com/dask/dask/issues/9969

It introduces a new `Task` class (name is subject to change) and a couple of other related subclasses that should replace the tuple as a representation of runnable tasks.

The benefits of this are outlined in https://github.com/dask/dask/issues/9969 but are primarily focused to reduce overhead during serialization and parsing of results. An important result is also that we can trivially cache functions (and arguments if we wish) to avoid problems like https://github.com/dask/distributed/issues/8767 where users are erroneously providing expensive to pickle functions (which also happens frequently in our own code and/or downstream projects like xarray)

This approach allows us to convert the legacy dsk graph to the new representation with full backwards compatibility. Old graphs can be migrated and new ones written directly using this new representation which will ultimately reduce overhead.

I will follow up with measurements shortly.

Sibling PR in distributed XXX

### CODE REVIEW

File: dask/task_spec.py [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+
+    def match_args_with_dependencies(self, values):
+        if len(values) < len(self.dependencies):
+            raise RuntimeError("Not enough arguments provided")
+        if not self.executables and not self.dep_to_arg:
+            return self.argspec
+        new_argspec = list(self.argspec)
+        if self.executables:
+            for tix in self.executables:
+                new_argspec[tix] = new_argspec[tix](values)
```
Comment (fjetter) [inline dask/task_spec.py:239]: Note how this can be pretty much also used to replace `SubgraphCallable`

**Comment by fjetter** [comment]
[preliminary benchmark results](https://github.com/coiled/benchmarks/actions/runs/10161188571) are... interesting...

I'll have to double check

image


FWIW in micro benchmarks the conversion of the old to new graph layout was up to 50% slower than the current unpacking code. While that sounds horrible, a lot of secondary checks (e.g. building the dependencies mappings) come for free which more or less amortizes this on small scale benchmarks. This strongly depends on the content of the graph, e.g. SubgraphCallables (from fusion) are more expensive to parse than other constructs.

I'm surprised to see any strong movement on the benchmark and beyond wall time, memory usage also drastically changes. Most likely, I made a mistake with the environments ¯\_(ツ)_/¯

File: dask/task_spec.py [hunk 2]
```diff
@@ -0,0 +1,742 @@
...
+from dask.typing import Key as KeyType
+from dask.utils import funcname, is_namedtuple_instance
+
+if TYPE_CHECKING:
+    from typing import TypeVar
+
+    _T = TypeVar("_T", bound="BaseTask")
+
+
+class WrappedKey:
```
Comment (fjetter) [inline dask/task_spec.py:103]: This had to move from distributed since we'd get otherwise horrible circular imports

File: dask/task_spec.py [hunk 3]
```diff
@@ -0,0 +1,742 @@
...
+    new_dsk = {}
+    for k, arg in dsk.items():
+        t = convert_old_style_task(k, arg, all_keys, only_refs)
+        if isinstance(t, Alias) and t.key == k:
+            continue
+        new_dsk[k] = t
+    return new_dsk
+
+
+class KeyRef:
```
Comment (fjetter) [inline dask/task_spec.py:295]: TODO: Maybe this should inherit from WrappedKey or be replaced by it entirely.

File: dask/task_spec.py [hunk 4]
```diff
@@ -0,0 +1,742 @@
...
+        if not isinstance(value, LiteralTask):
+            return False
+        if self.key != value.key:
+            return False
+        if self.value != value.value:
+            return False
+        return True
+
+
+_auto = object()
```
Comment (fjetter) [inline dask/task_spec.py:425]: todo: unused

File: dask/task_spec.py [hunk 5]
```diff
@@ -0,0 +1,742 @@
...
+        "key",
+        [Task("key-1", func, "a"), Task("key-2", func, "b")]
+    )
+
+    {"a": func("b")} ~ DictOfTasks("key", {"a": Task("a", func, "b")})
+
+    "literal-string" ~ Literal("key", "literal-string")
+
+
+Keys, Aliases and KeyRefs
```
Comment (phofl) [inline dask/task_spec.py:25]: How would I rename a result that was previously stored under a different key?

``{"key": KeyRef("old-key")}``?

**Review by phofl — COMMENTED** [review]
gave this a first look and looks good!

File: dask/task_spec.py (hunk 5 above)
Comment (fjetter) [reply dask/task_spec.py:25]: yes, this

**Comment by fjetter** [comment]
In case anybody reviews. Just had a conversation with Hendrik and here are a couple of expected concerns that were raised and discussed

- Sequence and Dict of task are relatively useless and we may be better off killing them entirely in favor of an ordinary task that encodes the list/dict generation. During development this was a (kind of) helpful construct but I wouldn't expect this to be used by any user. User API is pretty much only Task and KeyRef
- LiteralTask is by itself not very useful but this maintains an abstraction where all values of a low level graph obey the same callable structure which makes it quite natural to recurse into it. Again, this wouldn't be used by end users (and likely not even devs outside of the parsing code)

File: dask/task_spec.py [hunk 6]
```diff
@@ -0,0 +1,905 @@
...
+            for t in self.tasks.values()
+        ):
+            return self
+        return LiteralTask(self.key, self())
+
+    def __sizeof__(self):
+        return super().__sizeof__() + sizeof(self.tasks)
+
+
+class SequenceOfTasks(BaseTask):
```
Comment (fjetter) [inline dask/task_spec.py:773]: Note that I opted to keep this after all. It is not something end users (or even developers) really have to touch but it makes internals of the Task class much smoother. I'm not married to it, though, and could see it being removed down the line again. It would just require a lot more iterating over things, checking for types, etc.

File: dask/task_spec.py
Comment (fjetter) [inline dask/task_spec.py:1]: There is an open question about public API surface.

I'm inclined to make this module private for now but export something like `Task` or `WrappedKey` top level but I don't have a strong opinion here.

**Review by fjetter — COMMENTED** [review]
Right now, the Task signature is `key: Key, func: Callable, args: tuple, kwargs: dict` but it is also possible to make it `key: Key, func: Callable, /, *args: Any, **kwargs: Any` which would make it more natural to write a task, e.g. `Task(key, func, arg1, arg2, kwarg1='foo')`.

I chose to not go down this path because it makes internals a little more complex (namely we'd have to either make SequenceOfTasks iterable or remove it entirely). The additional complexity is OK if we want this API change

File: dask/task_spec.py [hunk 7]
```diff
@@ -0,0 +1,905 @@
...
+    new_dsk = {}
+    if cache is None:
+        cache = {}
+    for k, arg in dsk.items():
+        t = convert_old_style_task(k, arg, all_keys, only_refs, cache=cache)
+        if not only_refs and isinstance(t, Alias) and t.key == k:
+            continue
+        new_dsk[k] = t
+    new_dsk2 = {
+        k: v if isinstance(v, BaseTask) else LiteralTask(k, v)
```
Comment (phofl) [inline dask/task_spec.py:310]: We can't avoid this second loop?

File: dask/task_spec.py [hunk 8]
```diff
@@ -0,0 +1,905 @@
...
+        k = work.pop()
+        if k in seen or k not in dsk:
+            continue
+        seen.add(k)
+        t = dsk[k]
+        if (
+            isinstance(t, Alias)
+            and t.key not in keys
+            and t.key != k
+            and t.key in dsk
```
Comment (phofl) [inline dask/task_spec.py:341]: t.key not being in dsk means that we access a persisted result for example?

File: dask/task_spec.py [hunk 9]
```diff
@@ -0,0 +1,905 @@
...
+            and len(dependents[t.key]) == 1
```
Comment (phofl) [inline dask/task_spec.py:342]: This is protection if t.key is actually an input for a non-trivial task?

We would also catch a key that is an alias for 2 different results?

**Review by phofl — COMMENTED** [review]
This change is only hooked into distributed? I.e. if we get an old style grah then the sync scheduler will still execute the old style graph?

Wouldn't we want to do https://github.com/dask/distributed/pull/8797/files#r1731514684 client side to reduce the upload size?

**Comment by fjetter** [comment]
> Wouldn't we want to do https://github.com/dask/distributed/pull/8797/files#r1731514684 client side to reduce the upload size?

That would force materialization client side as well... which may be ok... I have to think about this a little more. 

It may be the case that we're currently always materializing things. Arrays are materialized because low level fusion is still enabled and dataframes are materialized because they are using dask-expr. I guess the only case where this is not true are some trivial things like `Client.map`.

File: dask/core.py [hunk 10]
```diff
@@ -189,6 +197,8 @@ def keys_in_tasks(keys: Collection[Key], tasks: Iterable[Any], as_list: bool = F
                 work.extend(w)
             elif typ is dict:
                 work.extend(w.values())
+            elif isinstance(w, BaseTask):
```
[the remark this answers is not in the document]
Comment (fjetter) [reply dask/core.py:200]: `isinstance` is relatively expensive compared to the other operations and as it stands right now, `BaseTask`s are still much rarer than old style tasks

File: dask/task_spec.py [hunk 11]
```diff
@@ -0,0 +1,905 @@
...
+        return SequenceOfTasks(None, obj).propagate_literal()
+
+    if isinstance(obj, (WrappedKey, KeyRef)):
+        return Alias(obj.key)
+
+    return obj
+
+
+def _wrap_namedtuple_task(k, obj, parser):
+    if hasattr(obj, "__getnewargs_ex__"):
```
[the remark this answers is not in the document]
Comment (fjetter) [reply dask/task_spec.py:184]: https://docs.python.org/3/library/pickle.html#pickling-class-instances

File: dask/task_spec.py [hunk 12]
```diff
@@ -0,0 +1,905 @@
...
+        else:
+            new_args = map(inner_args, args)
+            if only_refs:
+                # this can't be a literal since the literal wouldn't execute
+                # new_args in case there were runnable tasks or other literals
+                # See test_convert_task_only_ref
+                return Task(k, identity, (func, *new_args))
+            return Task(key=k, func=func, args=tuple(new_args))
+    try:
+        if isinstance(arg, (bytes, int, float, str, tuple)) and arg in all_keys:
```
[the remark this answers is not in the document]
Comment (fjetter) [reply dask/task_spec.py:265]: Theoretically, yes. This is a remnant of python2 compat where `str` was of type bytes and not unicode.

We should change this but I never put in the time to clean everything up. FWIW these two things needed to change

https://github.com/dask/dask/blob/2fbe18b29378f281f87341253c743f98f8e9ac72/dask/core.py#L202-L217
https://github.com/dask/dask/blob/2fbe18b29378f281f87341253c743f98f8e9ac72/dask/typing.py#L28

(and whatever pops up with mypy then).

I haven't used iskey here for perf reasons

File: dask/task_spec.py [hunk 13]
```diff
@@ -0,0 +1,905 @@
...
+
+        return DictOfTasks(None, new_dsk).propagate_literal()
+
+    elif isinstance(arg, WrappedKey):
+        return Alias(arg.key)
+    else:
+        return arg
+
+
+def convert_old_style_dsk(
```
[the remark this answers is not in the document]
Comment (fjetter) [reply dask/task_spec.py:293]: How about `convert_legacy_dsk` and `convert_legacy_task`?

File: dask/task_spec.py (hunk 7 above)
Comment (fjetter) [reply dask/task_spec.py:310]: perf wise that doesn't really matter (the only additional work we are doing here is initialize the for loop, instantiating the iterator, etc.)
The cost for the loop body is the same, i.e. we'll have to do the isinstance check for every task regardless of whether this is another loop or not.

The reason why this is here in the first place is because this ensures that after the conversion layer, every value in `dsk` is of type `BaseTask` which is not ensured by `convert_old_style_task`. I chose to not do this conversion in `convert_old_style_task` right away since `convert_old_style_task` is used recursively which would force us to wrap *every* object we encounter (which also works, this was my initial version but this is obviously slower)

File: dask/task_spec.py (hunk 8 above)
Comment (fjetter) [reply dask/task_spec.py:341]: yes

File: dask/task_spec.py (hunk 9 above)
Comment (fjetter) [reply dask/task_spec.py:342]: This behavior is tested in `test_resolve_aliases`

```python
dsk = {
        "bar": "foo",
        "foo": (func, "a", "b"),
        "baz": "bar",
        "foo2": (func, "bar", "c"),
    }
```

will convert to

```python
{
    'bar': Task('foo'), 
    'baz': Alias(bar), 
    'foo2': Task('foo2')
}
```

i.e. `bar` is pointing to `foo` and `foo` is only used by `bar`. Therefore, `bar` will just inherit the task (Now that I put this here, I notice that we should likely change the key of that Task object as well)

However, `bar` is being used in `baz` and `foo2` so inlining it would require us to compute it twice.

Comment (phofl) [reply dask/task_spec.py:341]: Can you add a short comment, it took me a while to remember what exactly this was supposed to catch

File: dask/task_spec.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (phofl) [reply dask/task_spec.py:292]: I like those better, yes

File: dask/task_spec.py (hunk 7 above)
Comment (phofl) [reply dask/task_spec.py:309]: Thanks!

File: dask/task_spec.py (hunk 8 above)
Comment (phofl) [reply dask/task_spec.py:340]: Can you add a brief comment for this as well? Took me a while to come to that conclusion

**Comment by phofl** [comment]
> That would force materialization client side as well... which may be ok... I have to think about this a little more.

Yeah I think maybe we only do this if the graph is materialised already? Not sure if this would add a lot of complexity though

File: dask/core.py [hunk 14]
```diff
@@ -43,6 +43,12 @@ def istask(x):
     >>> istask(1)
     False
     """
+    from dask.task_spec import BaseTask, LiteralTask
+
+    if isinstance(x, BaseTask):
+        if isinstance(x, LiteralTask):
+            return False
+        return True
     return type(x) is tuple and x and callable(x[0])
```
Comment (hendrikmakait) [inline dask/core.py:52]: It feels like this nested if-clause points out an inconsistency in the API (naming). x is a task if it's if's not a literal task. It could be that the ideal solution here would be renaming `istask` to something like `is_runnable_task` which I think I don't want to do here, but I wanted to point out that this feels weird.

File: dask/optimization.py [hunk 15]
```diff
@@ -345,17 +346,23 @@ def inline_functions(
...
-            return False
-
-    keys = [
-        k
-        for k, v in dsk.items()
-        if istask(v) and dependents[k] and k not in output and inlinable(v)
-    ]
+    def inlinable(key, task):
+        if (
+            not isinstance(task, BaseTask)
```
Comment (hendrikmakait) [inline dask/optimization.py:351]: Are `Tasks` never inlineable? Is this because we inline elsewhere?

File: dask/task_spec.py (hunk 3 above)
Comment (hendrikmakait) [reply dask/task_spec.py:354]: For me, calling this class `TaskRef` would make more sense since it does reference a `Task` via its key.
```suggestion
class TaskRef:
```

File: dask/task_spec.py [hunk 16]
```diff
@@ -0,0 +1,912 @@
...
+    while work:
+        k = work.pop()
+        if k in seen or k not in dsk:
+            continue
+        seen.add(k)
+        t = dsk[k]
+        if (
+            isinstance(t, Alias)
+            and t.key not in keys
+            and t.key != k
```
Comment (hendrikmakait) [inline dask/task_spec.py:339]: What would actually happen in this self-referential case?

File: dask/task_spec.py (hunk 6 above)
Comment (hendrikmakait) [reply dask/task_spec.py:777]: I'm not necessarily against these classes, I mostly dislike their name which suggests less magic than they actually perform. This might be rather something like a `Sequence{Consolidation|Conflation|Combination|Comprehension|Reduction}Task`

**Comment by fjetter** [comment]
I added a commit that changed the signature to `Task(key, func, /, *args, **kwargs)` which makes internals sometimes a little more awkward but is much more intuitive to use, I think

File: dask/_task_spec.py [hunk 17]
```diff
@@ -0,0 +1,928 @@
...
+
+
+_no_deps: set = set()
+
+
+class Alias(BaseTask):
+    __weakref__: Any = None
+    __slots__ = tuple(__annotations__)
+
+    def __init__(self, key: KeyRef | KeyType):
```
Comment (hendrikmakait) [inline dask/_task_spec.py:422]: The `Alias` class feels slightly off. It's the only `BaseTask` class that does not return the key with which it's also registered - or at least, should be registered with - into `dsk`. Instead, it returns the aliased key. I think that the intention is to make it simple to skip the Alias, but we have to implement code for optimizing `Alias`es away nonetheless since users could also just explicitly use the key from `dsk`. This might be a pain we're facing because the existing API forces us into allowing `Alias`es. A more explicit variant would be something like `Alias(old, new)`

Comment (fjetter) [reply dask/_task_spec.py:422]: An earlier version distinguished those and I had very often `Alias(key, key)` objects which felt a little redundant but I think I can make it work. It also struck me as weird to have the `key` attribute allowed to be different.

`Alias(key, ref=None)` seems appropriate

File: dask/optimization.py (hunk 15 above)
Comment (fjetter) [reply dask/optimization.py:351]: this is because legacy and new don't play well with each other. 

The `convert_legacy_*` can handle a mix of tasks but other layers cannot (so far, only distributed-only graphs come with `Task` objects to the conversion in distributed alone is sufficient. The very next step here should be to do a similar change in dask/dask to allow internals to assume we're working in the new system.


TLDR for now, legacy low level optimization is/should be disabled whenever a BaseTask is encountered.

File: dask/task_spec.py (hunk 16 above)
Comment (fjetter) [reply dask/task_spec.py]: at the very least dask.order raises a "cycles found" exception because strictly speaking a graph with this kind of self reference is not a DAG.

We had similar logic in the distributed scheduler before so without this kind of filtering, nothing will work. It's just a question about where we do this

File: dask/task_spec.py (hunk 3 above)
Comment (fjetter) [reply dask/task_spec.py]: I'm actually leaning towards nuking this class entirely in favor of `WrappedKey` (keeping that for backwards compat) but I'm happy to have `TaskRef` as an alias

File: dask/_task_spec.py (hunk 17 above)
Comment (fjetter) [reply dask/_task_spec.py]: I refactored this in https://github.com/dask/dask/pull/11248/commits/60df5d92356edec427093209b9cf45d9fafca992

I used `target` for the variable name instead of ref or new (since new is not always true, sometimes this is a genuine reference and ref is already taken)

Comment (fjetter) [reply dask/_task_spec.py]: This also makes the `resolve_aliases` (function name is not ideal) code a little easier to read imo

Comment (hendrikmakait) [reply dask/_task_spec.py]: Sounds good, `target` definitely works. I wanted to avoid something ambiguous like `Alias(key, alias)` where it's unclear what the newly introduced name is.

File: dask/_task_spec.py [hunk 18]
```diff
@@ -0,0 +1,941 @@
...
+        if isinstance(value, KeyRef):
+            return self.key == value.key
+        return self.key == value
+
+    def __hash__(self) -> int:
+        return hash(self.key)
+
+
+_func_cache: MutableMapping = {}
+_func_cache_reverse: MutableMapping = {}
```
Comment (fjetter) [inline dask/_task_spec.py:389]: Note this cache is unbounded. I didn't want to use an LRU (also because the acutally usable class for this is in distributed and the functools cache, etc. is not good). Also it's unclear how large the LRU should be.

Ideally, this was a cache that would work on bytes size so we could use zict, etc. for this but I didn't want to deal with this complexity just yet.

File: dask/_task_spec.py [hunk 19]
```diff
@@ -0,0 +1,924 @@
...
+            continue
+        seen.add(k)
+        t = dsk[k]
+        if isinstance(t, Alias):
+            target_key = t.target.key
+            # Rules for when we allow to collapse an alias
+            # 1. The target key is not in the keys set. The keys set is what the
+            #    user is requesting and by collapsing we'd no longer be able to
+            #    return that result.
+            # 2. The target key is in fact part of dsk. If it isnt' this could
```
Comment (hendrikmakait) [inline dask/_task_spec.py:321]:
```suggestion
            # 2. The target key is in fact part of dsk. If it isn't this could
```

File: dask/core.py (hunk 14 above)
Comment (hendrikmakait) [reply dask/core.py:52]: Agreed!

FWIW, I like `GraphNode`, `DataLiteral` (or even just `Data`/`Datum`), and `Task`. Feel _very_ free not to rename things right now though.

File: dask/base.py [hunk 20]
```diff
@@ -1095,7 +1095,9 @@ def _seen_ctx(reset: bool) -> Iterator[dict[int, tuple[int, object]]]:
             # tokenize()
             seen = {}
             tok = _seen.set(seen)
-
+        else:
+            seen = seen.copy()
+            tok = _seen.set(seen)
```
Comment (fjetter) [inline dask/base.py:1100]: So, I'm running into an issue on dask/distributed in the test `test_threadsafe_get`. The tests is launching 30 threads that are concurrently hammering the scheduler with the same computation request over and over again, i.e. the tokens should always be identical and the result should in most likeliness even be reused between all the threads.
However, I'm running into problems and this appears to fix it. Haven't managed to isolate it (already worked with RLocks in various places which didn't help). The problem seems to be a counting issue in the `seen` dictionary but I couldn't reduce it or write a test for it.

I'm slightly concerned about this fix since we call this function quite a lot and copying a non-empty set can be expensive (haven't done any measurements, yet)

File: dask/_task_spec.py [hunk 21]
```diff
@@ -0,0 +1,937 @@
...
+        return self
+
+    def copy(self):
+        return DataNode(self.key, self.value)
+
+    def __call__(self, values=()):
+        return self.value
+
+    def __repr__(self):
+        return f"Literal({self.key}, type={self.typ}, {self.value})"
```
Comment (hendrikmakait) [inline dask/_task_spec.py:501]:
```suggestion
        return f"DataNode({self.key}, type={self.typ}, {self.value})"
```

File: dask/optimization.py [hunk 22]
```diff
@@ -684,6 +696,11 @@ def fuse(
                         and
                         # Sanity check; don't go too deep if new levels introduce new edge dependencies
                         (no_new_edges or height < max_depth_new_edges)
+                        and (
+                            not isinstance(dsk[parent], GraphNode)
+                            # TODO: substitute can be implemented with GraphNode.inline
```
Comment (hendrikmakait) [inline dask/optimization.py:701]: Should we add an issue for this?

**Review by hendrikmakait — APPROVED** [review]
Thanks, @fjetter! This looks overall good to me. I have a few non-blocking comments regarding consistent wording and adding issues for remaining TODOs in the code.

**Comment by fjetter** [comment]
FYI I'm currently working on removing the Sequence/Dict classes. I suspect there could be a perf gain. I'll report back once it is stable. Either way, we can move forward with this. Without any further changes in distributed or in this repo, this cahnge alone is pretty harmless

**Comment by fjetter** [comment]
I have a follow up ready that addresses all of the above review comments and removes the sequence/dict tasks. In favor of having a simple review, I'll move forward with merging this PR (will not affect anything, yet)
